libcxxwrap_julia: switch to libjulia

fingolfin opened the pull request:
Also drop the restriction to the cxx11 ABI and add more platforms.

Motivation: while this is not the perfect next step for libcxxwrap_julia, I think it makes sense to do this now: switching to libjulia and adding more platforms will allow other packages using libcxxwrap_julia to do the same. Once (if?) we get "full" support for "julia_version" as part of the platform, adjusting `libcxxwrap_julia` and other JLLs depending on it to benefit from those shouldn't be hard; but it is unclear when this will be available; so why not implement those benefits we can already get now?

As a side effect, this will produce a fresh JLLWrapper based JLL :-)

However, this PR shouldn't be merged without @barche approving.

UPDATE: ... and of course it should pass the CI tests before all else ;-)

Close #1647

fingolfin commented:
So there are still issues with using libjulia here -- so I think it's good indeed to try this now, debugging this might take enough time for someone to finish `julia_version` platform support 😂 

For starters, I am seeing this:
```
-- Julia_LIBRARY:        Julia_LIBRARY-NOTFOUND
CMake Error at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:164 (message):
  Julia not found (missing: Julia_LIBRARY Julia_LIBRARY_DIR) (found version
  "1.4.2")
Call Stack (most recent call first):
  /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:445 (_FPHSA_FAILURE_MESSAGE)
  FindJulia.cmake:173 (find_package_handle_standard_args)
  CMakeLists.txt:17 (find_package)
```
and also
```
CMake Error at FindJulia.cmake:32 (file):
  file failed to open for reading (No such file or directory):

    /workspace/srcdir/libcxxwrap-julia/Julia_VERSION_INCLUDE-NOTFOUND
Call Stack (most recent call first):
  CMakeLists.txt:17 (find_package)


CMake Error at FindJulia.cmake:33 (string):
  string sub-command REGEX, mode MATCH needs at least 5 arguments total to
  command.
Call Stack (most recent call first):
  CMakeLists.txt:17 (find_package)
```

So that suggests that `FindJulia.cmake` and/or `libjulia` may need work.

And then there is also this:
```
/opt/bin/i686-linux-gnu-g++ --sysroot=/opt/i686-linux-gnu/i686-linux-gnu/sys-root/  -DJLCXX_EXPORTS -DJULIA_ENABLE_THREADING -Dcxxwrap_julia_EXPORTS -I/workspace/srcdir/libcxxwrap-julia/include -I/workspace/destdir/include/julia  -Wunused-parameter -Wextra -Wreorder -O3 -DNDEBUG -fPIC   -std=gnu++1z -o CMakeFiles/cxxwrap_julia.dir/src/jlcxx.cpp.o -c /workspace/srcdir/libcxxwrap-julia/src/jlcxx.cpp
In file included from /workspace/destdir/include/julia/julia.h:73:0,
                 from /workspace/srcdir/libcxxwrap-julia/include/jlcxx/type_conversion.hpp:4,
                 from /workspace/srcdir/libcxxwrap-julia/include/jlcxx/array.hpp:4,
                 from /workspace/srcdir/libcxxwrap-julia/src/c_interface.cpp:1:
/workspace/destdir/include/julia/atomics.h:9:4: error: #error Julia can only be built for architectures above Pentium 4. Pass -march=pentium4, or set MARCH=pentium4 and ensure that -march is not passed separately with an older architecture.
 #  error Julia can only be built for architectures above Pentium 4. Pass -march=pentium4, or set MARCH=pentium4 and ensure that -march is not passed separately with an older architecture.
    ^~~~~
```

benlorenz commented:
To fix mingw please try adding `-DJulia_LIBRARY=$libdir/libjulia.$dlext` (it shouldn't hurt setting this for other targets as well).
(FindJulia expects a `libjulia.dll.a` in `$prefix/lib` but libjulia only has the shared library `libjulia.dll` in `$libdir == $prefix/bin`)

And there is no libjulia_jll for freebsd (yet?): https://github.com/JuliaPackaging/Yggdrasil/blob/fddff44e98287fc7ce76118b95ffa4dadcfbc6a6/L/libjulia/common.jl#L180

fingolfin commented:
Thanks, I applied these suggestions. Unfortunately, there are still failures, but related to issues with MPFR_jll v4.0.2+4. That's really mysterious?!?
```
┌ Warning: Dependency MPFR_jll does not have an Artifacts.toml at /depot/packages/MPFR_jll/735jI/Artifacts.toml!
└ @ BinaryBuilderBase /depot/packages/BinaryBuilderBase/j5xoW/src/Prefix.jl:470
##[warning]Dependency MPFR_jll does not have an Artifacts.toml at /depot/packages/MPFR_jll/735jI/Artifacts.toml!
```

fingolfin commented:
This seems to work now! Thanks to @benlorenz for the help.

@barche care to take a look?

barche commented:
This looks great to me, thanks for doing this. Just two doubts:
- I don't like the need to set `-DJulia_LIBRARY` manually, this will need to be done on all dependent packages. Should we fix this in FindJulia directly?
- I'd prefer to release a version based on libjulia1.3 and then the 1.4 one

fingolfin commented:
Let's wait for PR #1951 and its two follow-ups, then I'll redo this PR .

fingolfin commented:
Updated for that latest libjulia_jll. Unfortunately there is a compile error on FreeBSD: `/workspace/destdir/include/julia/ios.h:7:10: fatal error: 'uv.h' file not found`. But I guess we could go without FreeBSD support for now...

Regarding `Julia_LIBRARY` -- yeah it would be good if that could be fixed centrally, instead of forcing all other JLLs using libcxxwrap to adapt.

giordano commented:
> Updated for that latest libjulia_jll. Unfortunately there is a compile error on FreeBSD: `/workspace/destdir/include/julia/ios.h:7:10: fatal error: 'uv.h' file not found`. But I guess we could go without FreeBSD support for now...

We could go without FreeBSD (as long as @ararslan doesn't see us), but the error is rather weird :thinking:

fingolfin commented on L/libcxxwrap_julia/build_tarballs.jl:
```diff
@@ -18,23 +19,36 @@ sources = [
+# workaround for 32bit intel builds, where Julia's atomics.h will otherwise
+# complain
+if [[ "${target}" == i686* ]]; then
+    export CXXFLAGS="-msse -msse2"
```
I really don't like this bit. I feel that it should be up to libjulia_jll to ensure that projects using it gets suitable compiler flags, somehow. As it is, we'll likely have to put this into all JLL build recipes which are using libjulia_jll, and that doesn't scale well.

How to do that? Perhaps teach `julia-config.jl --cflags` (by patching it) to add these? But is libcxxwrap even using this? If not, perhaps it should?

Alternatively, perhaps libcxxwrap_julia could extend its CMake macros to insert these compilers flags, then at least it and presumably all (?) JLLs using it would be OK?

giordano replied:
I hadn't found the time to investigate this, but it looks weird that we have to specify those features, since those should be default for the target platform

fingolfin replied:
```
        "i686" => Dict(
            "i686" => ["-march=i686", "-mtune=generic"],
            "prescott" => ["-march=prescott", "-mtune=prescott"],
        ),
```
And here `i686` is used which doesn't enable SSE by default. That would be the case with `prescott`, though. 

So I am going out on a limb here, as admittedly I have not understood this part of BB at all, but: Perhaps the problem then is that libjulia_jll is built for i686 instead of prescott? I think BB tries to "guess" this, but for the x86_64 builds of libjulia, I see this warning:
> ##[warning]libjulia.so.1.6 contains a `cpuid` instruction; refusing to analyze for minimum instruction set, as it may dynamically select the proper instruction set internally.  Would have chosen avx2, instead choosing x86_64.

Now to my surprise I see no similar warning for the i686 builds? Could it perhaps be tha the code detecting the minimal march is incomplete and/or buggy?

fingolfin replied:
Aha, from `minimum_march` in `BinaryBuilder.jl`:
```
    elseif arch(p) == "i686"
        if counts["sse3"] > 0
            return "prescott"
        end
```
But the only alternative right now is to use `i686` as arch, which means pentium pro. To get SSE and SSE2, you'd want at least pentium4.

So either another arch ("pentium4") needs to be added, or else the prescott check be extended to also cover MMX, SSE and SSE2.

fingolfin replied:
Third option: changing `i686` to actually map to `pentium4` (given that Julia won't run on anything older, so why bother restricting JLL binaries further than that? Might as well benefit from additional optimizations, too.

fingolfin commented:
`uv.h` should come from `LibUV_jll`, and I just verified that it is in there, including the FreeBSD variant.

But I guess the compiler might not pick up the include path where `uv.h` has been installed to? But then why does it work on other platforms?

barche commented:
> Regarding `Julia_LIBRARY` -- yeah it would be good if that could be fixed centrally, instead of forcing all other JLLs using libcxxwrap to adapt.

With commit `2ee08fa0877f7268d06fd23c1be1b7c80b5de48c` on libcxxwrap-julia it should work without the `Julia_LIBRARY`. Best to also bump the version to v0.8.2, which I will tag if this works out OK.

benlorenz commented:
> `uv.h` should come from `LibUV_jll`, and I just verified that it is in there, including the FreeBSD variant.
> 
> But I guess the compiler might not pick up the include path where `uv.h` has been installed to? But then why does it work on other platforms?

Probably because on FreeBSD (in general and not just in BB) `<sysroot>/usr/local/include` is not in the default path when searching headers and `$destdir` (containing all the dependencies) is linked to `<sysroot>/usr/local`.
I guess it would make sense to add that directory as a search path in the compiler wrappers for freebsd?

Otherwise adding `-I$includedir` would also help but then needs to be replicated in all other related build scripts.

```
sandbox:${WORKSPACE}/srcdir/build # /opt/bin/clang++ --sysroot=/opt/x86_64-unknown-freebsd11.1/x86_64-unknown-freebsd11.1/sys-root/  -DJLCXX_EXPORTS -DJULIA_ENABLE_THREADING -Dcxxwrap_julia_EXPORTS -I/workspace/srcdir/libcxxwrap-julia/include -I/workspace/destdir/include/julia  -Wunused-parameter -Wextra -Wreorder -O3 -DNDEBUG -fPIC   -std=gnu++17 -o CMakeFiles/cxxwrap_julia.dir/src/jlcxx.cpp.o -c /workspace/srcdir/libcxxwrap-julia/src/jlcxx.cpp -v
clang version 9.0.1 (/home/staticfloat/.julia/dev/BinaryBuilder/deps/downloads/llvm-project.git c1a0a213378a458fbea1a5c77b315c7dce08fd05)
Target: x86_64-unknown-freebsd11.1
Thread model: posix
InstalledDir: /opt/x86_64-linux-musl/bin
 "/opt/x86_64-linux-musl/bin/clang-9" -cc1 -triple x86_64-unknown-freebsd11.1 -aux-triple x86_64-unknown-freebsd11.1 -emit-obj -disable-free -disable-llvm-verifier -discard-value-names -main-file-name jlcxx.cpp -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -masm-verbose -mconstructor-aliases -munwind-tables -target-cpu x86-64 -dwarf-column-info -debugger-tuning=gdb -v -coverage-notes-file /workspace/srcdir/build/CMakeFiles/cxxwrap_julia.dir/src/jlcxx.cpp.gcno -resource-dir /opt/x86_64-linux-musl/lib/clang/9.0.1 -D JLCXX_EXPORTS -D JULIA_ENABLE_THREADING -D cxxwrap_julia_EXPORTS -I /workspace/srcdir/libcxxwrap-julia/include -I /workspace/destdir/include/julia -D NDEBUG -isysroot /opt/x86_64-unknown-freebsd11.1/x86_64-unknown-freebsd11.1/sys-root/ -internal-isystem /opt/x86_64-unknown-freebsd11.1/x86_64-unknown-freebsd11.1/sys-root//usr/include/c++/v1 -O3 -Wunused-parameter -Wextra -Wreorder -std=gnu++17 -fdeprecated-macro -fdebug-compilation-dir /workspace/srcdir/build -ferror-limit 19 -fmessage-length 97 -fobjc-runtime=gnustep -fcxx-exceptions -fexceptions -fdiagnostics-show-option -fcolor-diagnostics -vectorize-loops -vectorize-slp -fopenmp-is-device -faddrsig -o CMakeFiles/cxxwrap_julia.dir/src/jlcxx.cpp.o -x c++ /workspace/srcdir/libcxxwrap-julia/src/jlcxx.cpp
clang -cc1 version 9.0.1 based upon LLVM 9.0.1 default target x86_64-linux-musl
#include "..." search starts here:
#include <...> search starts here:
 /workspace/srcdir/libcxxwrap-julia/include
 /workspace/destdir/include/julia
 /opt/x86_64-unknown-freebsd11.1/x86_64-unknown-freebsd11.1/sys-root//usr/include/c++/v1
 /opt/x86_64-linux-musl/lib/clang/9.0.1/include
 /opt/x86_64-unknown-freebsd11.1/x86_64-unknown-freebsd11.1/sys-root//usr/include
End of search list.
In file included from /workspace/srcdir/libcxxwrap-julia/src/jlcxx.cpp:1:
In file included from /workspace/srcdir/libcxxwrap-julia/include/jlcxx/array.hpp:4:
In file included from /workspace/srcdir/libcxxwrap-julia/include/jlcxx/type_conversion.hpp:4:
In file included from /workspace/destdir/include/julia/julia.h:12:
In file included from /workspace/destdir/include/julia/libsupport.h:13:
/workspace/destdir/include/julia/ios.h:7:10: fatal error: 'uv.h' file not found
#include "uv.h"
         ^~~~~~
1 error generated.
```

fingolfin commented:
Huh, OK, interesting. So how come other packages which rely on headers from JLLs installed as (build) dependency do compile on FreeBSD? Are they all already adding in explicit `-I` flags?

fingolfin commented:
> With commit `2ee08fa0877f7268d06fd23c1be1b7c80b5de48c` on libcxxwrap-julia it should work without the `Julia_LIBRARY`. Best to also bump the version to v0.8.2, which I will tag if this works out OK.

@barche I switched to that commit and it seems to work as intended

benlorenz commented:
> Huh, OK, interesting. So how come other packages which rely on headers from JLLs installed as (build) dependency do compile on FreeBSD? Are they all already adding in explicit `-I` flags?

In some cases cmake or configure might determine that these are needed, so this could probably be fixed in FindJulia as well. But it is also given in the documentation that it might be needed in some cases (a quick grep found about 100 such flags in the recipes): https://juliapackaging.github.io/BinaryBuilder.jl/dev/troubleshooting/#Header-files-of-the-dependencies-can't-be-found
For `LDFLAGS` there is even a comment that this was needed just for FreeBSD in one case.

In libjulia there is also `CFLAGS="-I${prefix}/include"` and `LDFLAGS="-L${prefix}/lib"`.

barche commented:
I hope it's not rude to push changes, but I think with the latest update of FindJulia it will work. Fingers crossed.

fingolfin commented:
Not rude at all, to the contrary; in fact, I was somewhat worried when I made this that you'd find me rude for doing a PR for "your" JLL... but I figured it'd be useful to get this kicked off to find obstacles and remove them, and, well, that's exactly what happened ;-). In fact, I am somewhat worried whether the JLL this PR produces works in practice -- so I think before this PR is merged, we really should test it locally on some systems (and even before that, perhaps download the tarballs created by the CI runners and inspect the binaries in them).

benlorenz commented on L/libcxxwrap_julia/build_tarballs.jl:
```diff
@@ -3,38 +3,45 @@
+# skip PowerPC builds in Julia 1.3 (not supported by libjulia_jll)
+if julia_version < v"1.4"
+    filter!(p -> !(Sys.islinux(p) && arch(p) == "powerpc64le"), platforms)
+end
```
```suggestion
end

platforms = expand_cxxstring_abis(platforms)
```
I think the abi expansion is missing by accident?

benlorenz commented:
I did a first test with the `x86_64-linux-gnu.tar.gz` tarball and my julia 1.5 installation: building the latest `libpolymake-julia` using this tree worked fine and also the Polymake.jl testsuite had no errors (using this tarball via `Overrides.toml`, i.e. using the old wrappers but the new binaries).

fingolfin commented:
Since this seems to work well now, how shall we proceed? Merge this, then @barche proceeds to make Julia 1.4 and 1.5 variants?

barche commented:
> Since this seems to work well now, how shall we proceed? Merge this, then @barche proceeds to make Julia 1.4 and 1.5 variants?

Yes, that's the idea, but this also reminds me: wasn't there a way to limit the Julia compat in the JLL? It would be better to have this done in the JLL rather than in CxxWrap.jl. In that case, the limit for this JLL should be Julia 1.3, but I forgot how this is done.

fingolfin commented:
Using `julia_compat`, you can grep for it.

But! there is a caveat: packages with the the same version, just different build version (the `+3` bit at the end of JLL versions) can't really have different compat restrictions. So you'll need to use different versions for this to work, even if they only differ in the patch level... might be a good idea to check back with @giordano and @staticfloat , too

fingolfin commented:
Perhaps one could abuse the prerelease part of the version string? I.e. use versions `0.8.3-julia.1.3`, `0.8.3-julia.1.4`, ... ?

fingolfin commented:
I've rebased and squashed this.

I've also added a `julia_compat` field and added the version number tweak (abusing the "prerelease" feature of semver). So in principle, we should now be able to build separate versions for Julia 1.3, 1.4, 1.5 and 1.6 (though that'll require #1987 to be merged).
Alternatively, one could hope that e.g. the 1.5 build will be compatible with 1.6 and later builds, and drop its compat restriction / change it to ">= 1.5". But that seems risky for a minuscule gain (one less JLL version to be build), yet requires extra work (tweaking the julia compat just for the 1.5 build), which in turn increases the risk of making a mistake when building this.

Another thing: We could split this recipe now into multiple (like we did for e.g. libjulia), that coexist; that might make maintenance a tad easier in the future? (Or not, dunno)

fingolfin commented:
huh, I am getting this on Yggdrasil, which I cannot reproduce locally:
```
Generating meta.json...
ERROR: LoadError: LoadError: ArgumentError: invalid version string: 0.8.2-julia.$(julia_version.major).$(julia_version.minor)
Stacktrace:
 [1] parse
   @ ./version.jl:126 [inlined]
 [2] VersionNumber(v::String)
   @ Base ./version.jl:130
 [3] var"@v_str"(__source__::LineNumberNode, __module__::Module, v::Any)
   @ Base version.jl:146
in expression starting at /agent/_work/1/s/L/libcxxwrap_julia/build_tarballs.jl:8
in expression starting at /agent/_work/1/s/L/libcxxwrap_julia/build_tarballs.jl:8
```

(Note that I am testing with Julia .6.0-DEV.1390 but Yggdrasil has 1.6.0-DEV.1275)

fingolfin commented:
OK, all working well now :-).

barche commented:
This looks good to me!

giordano commented on L/libcxxwrap_julia/build_tarballs.jl:
```diff
@@ -2,39 +2,49 @@
+
 name = "libcxxwrap_julia"
-version = v"0.8.0"
+version = VersionNumber("0.8.2-julia.$(julia_version.major).$(julia_version.minor)")
```
What's the rationale of this version number?  Pkg doesn't really like anything outside of X.Y.Z and we overstress it with the build number.

fingolfin replied:
Here's the reason: We need to build separate binaries (artifacts) of libcxxwrap for *at least Julia 1.3 versus Julia >= 1.4; but better, for each of Julia 1.3, 1.4, 1.5, 1.6.  Each of the corresponding JLL wrappers needs a different `compat` entry for Julia in `Project.toml`. But to do that, they need actual different versions (the build number part after the `+` is not enough for that, the registry doesn't support that, which makes sense, as it is consistent with how semver defines build number).

Unlike build numbers, the pre-release is part of semver which is explicitly defined to indicate different versions, including how they should be compared with each other and other versions. It also is used by Julia itself As such, I would hope that existing Julia tooling correctly deals with it (if not, that'd be a bug in the semver support of that tool, which IMHO should be fixed). 

One drawback is that the semver spec says this "A pre-release version indicates that the version is unstable and might not satisfy  the intended compatibility requirements as denoted by its associated normal version." However, that should not matter for any tooling, it's just something about the semantics for users; and the only user of this JLL would be CxxWrap.jl, so there should be no issue here, I think.

barche replied:
I think this is OK, as long as it works. Not knowing about pre-releases, what I had in mind for this was:
* libcxxwrap 0.8.2 = Julia 1.3
* libcxxwrap 0.8.3 = Julia 1.4
* libcxxwrap 0.8.4 = Julia 1.5
* libcxxwrap 0.8.5 = Julia 1.6

Starting from Julia 1.6, I am hoping this is solved by the julia version being part of the artifact resolution, if I understood correctly?

giordano replied:
> Starting from Julia 1.6, I am hoping this is solved by the julia version being part of the artifact resolution, if I understood correctly?

Well, that would work for Julia v1.6+

fingolfin replied:
@barche if you are willing to go through that hassle, sure, why not (but perhaps shift all the libcxxwrap version up by one, then the last digits of the libcxxwrap and Julia version match, that might make things easier?).

Of course it also raises the question how you then would version a bugfix release? Perhaps by adding 10 to the patchlevel, so that Julia 0.8.13 = Julia 1.3,  etc., then 0.8.23 = Julia 1.3 etc?

fingolfin replied:
@dilumaluthge just clarified on Slack: "We don't allow prereleases in the General registry."

fingolfin replied:
@barche so, what's the plan now? Do you want to take over this PR, or make a new one? (Feel free to use any or all of what's in here, or also feel free to directly modify this PR as you like)

barche replied:
Sorry for the slow response. I propose to leave the version number at ` v"0.8.0"`, I don't really plan any bugfix releases since this is a temporary situation until Julia 1.6 is out, would that be acceptable?

fingolfin replied:
? A single version wouldn't allow to produce several binaries for several Julia versions, though? And hence wouldn't allow to resolve the crashes with Julia 1.3.

Also, the current release is 0.8.0, hence this should be at least 0.8.1.

It is also far from clear to me that this is a temporary situation which will go away with Julia 1.6 (plus, it will likely be months until then, won't it? : first off, that assumes everyone is ready to immediately drop support for all Julia versions once 1.6.0 is released. Secondly, it is not yet clear to me that Julia 1.6 will really resolve this: are all pieces already there? Even if, without verifying that it works as desired in practice, I'd be reluctant to rely on that (no offense intended to @staticfloat and anybody else  who may habe worked on this).

barche replied:
No, what I mean is keep the proposed mapping as:

* libcxxwrap 0.8.2 = Julia 1.3
* libcxxwrap 0.8.3 = Julia 1.4
* libcxxwrap 0.8.4 = Julia 1.5
* libcxxwrap 0.8.5 = Julia 1.6

If you think bugfixes are important for older Julia versions, we could do:

* libcxxwrap 0.8.2 = Julia 1.3
* libcxxwrap 0.9.0 = Julia 1.4
* libcxxwrap 0.10.0 = Julia 1.5
* libcxxwrap 0.11.0 = Julia 1.6

fingolfin replied:
As long as they are all using the same source, I'd suggest to keep them all at 0.8.x

fingolfin replied:
I've changed the version (back) to 0.8.2, as discussed here. Can this be merged now, or is anything still to be done? (I am not aware of anything, but...)

barche replied:
OK for me!
